packages/backend/src/shared/core/Result.ts

```diff
@@ -0,0 +1,91 @@
+export class Result<T> {
+  public isSuccess: boolean;
+  public isFailure: boolean;
+  public error?: T | string;
+  private _value?: T;
+
+  public constructor(isSuccess: boolean, error?: T | string, value?: T) {
+    if (isSuccess && error) {
+      throw new Error('InvalidOperation: A result cannot be successful and contain an error');
+    }
+    if (!isSuccess && !error) {
+      throw new Error('InvalidOperation: A failing result needs to contain an error message');
+    }
+
+    this.isSuccess = isSuccess;
+    this.isFailure = !isSuccess;
+    this.error = error;
+    this._value = value;
+
+    Object.freeze(this);
+  }
+
+  public getValue(): T {
+    if (!this.isSuccess || !this._value) {
```

Comment:
in case if `this._value` is 0 or false, this case will fail, right? then we should explicitly check for null and undefined
```suggestion
    if (!this.isSuccess || this._value !== null || this._value !== undefined) {
```